nomad/vault.go

```diff
@@ -539,25 +530,77 @@ func (v *vaultClient) renewalLoop() {
 	}
 }
 
+// nextBackoff returns the delay for the next auto renew interval, in seconds.
+// Returns negative value if past expiration
+//
+// It should increase the amount of backoff each time, with the following rules:
+//
+// * If token expired already despite earlier renewal attempts,
+//   back off for 1 minute + jitter
+// * If we have an existing authentication that is going to expire,
+// never back off more than half of the amount of time remaining
+// until expiration (with 5s floor)
+// * Never back off more than 30 seconds multiplied by a random
+// value between 1 and 2
+// * Use randomness so that many clients won't keep hitting Vault
+// at the same time
+func nextBackoff(backoff float64, expiry time.Time) float64 {
+	maxBackoff := time.Until(expiry) / 2
+
+	if maxBackoff < 0 {
+		// expiry passed
+		return 60 * (1.0 + rand.Float64())
+	}
+
+	switch {
+	case backoff >= 24:
+		backoff = 30
+	default:
+		backoff = backoff * 1.25
+	}
+
+	// Add randomness
+	backoff = backoff * (1.0 + rand.Float64())
+
+	if backoff > maxBackoff.Seconds() {
+		backoff = maxBackoff.Seconds()
+	}
+
+	if backoff < 5 {
+		backoff = 5
+	}
+
+	return backoff
+}
+
 // renew attempts to renew our Vault token. If the renewal fails, an error is
-// returned. This method updates the lastRenewed time
-func (v *vaultClient) renew() error {
+// returned.  The boolean indicates whether it's safe to attempt to renew again.
+// This method updates the currentExpiration time
+func (v *vaultClient) renew() (bool, error) {
```

Comment:
Just FYI we have a [RecoverableError](https://github.com/hashicorp/nomad/blob/60c6cb858669261473ee59119e8dcd04055a2eef/nomad/structs/structs.go#L8495-L8544) type we sometimes use in circumstances like this. Since the recoverable aspect is handled within this package and doesn't cross public API boundaries, I don't think there's a compelling reason to switch to it.